### PR TITLE
Initial commit for open inference protocol

### PR DESCRIPTION
**Pull request by yuzisun**
Related to https://github.com/kserve/kserve/issues/2663

- Add open inference protocol specification for REST and gRPC

### CODE REVIEW

File: specification/protocol/inference_rest.md [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    $metadata_server_error_response =
+    {
+      "error": $string
+    }
+
+* “error” : The descriptive message for the error.
+
+
+The per-model metadata endpoint provides information about a model. A model metadata request is made with an HTTP GET to a model metadata endpoint. In the corresponding response the HTTP body contains the [Model Metadata Response JSON Object](#model-metadata-response-json-object) or the [Model Metadata Response JSON Error Object](#model-metadata-response-json-error-object).
+The model name and (optionally) version must be available in the URL. If a version is not provided the server may choose a version based on its own policies or return an error.
```
Comment (tessapham) [inline specification/protocol/inference_rest.md:130]: Duplicate of below.

File: specification/protocol/inference_rest.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+* "name" : The name of the input tensor.
+* "shape" : The shape of the input tensor. Each dimension must be an
+  integer representable as an unsigned 64-bit integer value.
+* "datatype" : The data-type of the input tensor elements as defined
+  in [Tensor Data Types](#tensor-data-types).
+* "parameters" : An object containing zero or more parameters for this
+  input expressed as key/value pairs. See [Parameters](#parameters)
+  for more information.
+* “data”: The contents of the tensor. See [Tensor Data](#tensor-data)
```
Comment (tessapham) [inline specification/protocol/inference_rest.md:266]: Will there be additional parts that explain Tensor Data, Tensor Data Types, Platforms, etc.?

File: specification/protocol/inference_rest.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+| Model Metadata | The "model metadata" API is a per-model endpoint that returns details about the model passed in the path. | 
+| Server Ready | The “server ready” health API indicates if all the models are ready for inferencing. The “server ready” health API can be used directly to implement the Kubernetes readinessProbe |
+| Server Live | The “server live” health API indicates if the inference server is able to receive and respond to metadata and inference requests. The “server live” API can be used directly to implement the Kubernetes livenessProbe. |
+| Server Metadata | The "server metadata" API returns details describing the server. | 
+| Model Ready | The “model ready” health API indicates if a specific model is ready for inferencing. The model name and (optionally) version must be available in the URL. |
+
+### Health/Readiness/Liveness Probes
+
+The Model Readiness probe the question "Did the model download and is it able to serve requests?" and responds with the available model name(s). The Server Readiness/Liveness probes answer the question "Is my service and its infrastructure running, healthy, and able to receive and process requests?"
```
Comment (tessapham) [inline specification/protocol/inference_rest.md:41]: Should this question be `"Was the model successfully downloaded and loaded into the server?"`. I thought it's the server that serves requests.

File: specification/protocol/inference_rest.md (hunk 1 above)
Comment (yuzisun) [reply specification/protocol/inference_rest.md]: fixed

**Review by theofpa — APPROVED** [review]

File: specification/protocol/inference_rest.md (hunk 3 above)
Comment (yuzisun) [reply specification/protocol/inference_rest.md:42]: good point, the model is loaded into the memory so the server is able to invoke the model's inference API upon receiving the request

File: specification/protocol/open_inference_grpc.proto [hunk 4]
```diff
@@ -0,0 +1,326 @@
+// Copyright 2020 kubeflow.org.
```
[the remark this answers is not in the document]
Comment (yuzisun) [reply specification/protocol/open_inference_grpc.proto]: fixed

File: specification/protocol/inference_rest.md (hunk 2 above)
Comment (yuzisun) [reply specification/protocol/inference_rest.md]: Thanks for pointing out, I missed these two sections.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,25 @@
...
+enpowering the development of versatile client or benchmarking tools that can work with all
+supported serving runtimes.
+
+- The inference REST [specification](./specification/protocol/inference_rest.md)
+- The inference gRPC [specification](./specification/protocol/inference_grpc.md)
+
+## Adoptions
+- KServe [v2 inference protocol](https://kserve.github.io/website/master/modelserving/data_plane/v2_protocol/)
+- NVIDIA [Triton inference server protocol](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/customization_guide/inference_protocols.html)
+- Seldon [MLServer v2 inference protocol](https://docs.seldon.io/projects/seldon-core/en/latest/reference/apis/v2-protocol.html)
```
Comment (adriangonz) [inline README.md:15]: Could we also add Seldon Core V2 as a separate entry? We could point to this link: 

https://docs.seldon.io/projects/seldon-core/en/v2/contents/getting-started/#api-for-inference

**Pull request merged**